### PR TITLE
Refactor/Curp test utils

### LINKED ISSUE #290: [Refactor]: Combine curp common test utils
Currently in `curp/test_utils` and `curp/tests/common`, there are quite amount of code serve the same purpose. The first one is used for unit tests and the latter one is used for integration tests. I think we should combine them to reduce duplication.

I plan to refactor the code to use a single set of tests utils. After the refactor the curp should expose the test utils in public API, and can be used in both internal integration tests and madsim simulation tests.

### PR DESCRIPTION
**Pull request by bsbds**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

    Closes #290 

* what changes does this pull request make?

    1. move curp public trait interfaces to `curp-extapi` crate (because of circular depenency, the curp-testutils cannot deps on curp directly, so make it deps on `curp-extapi`)

    2. move curp test utils to `curp-testutils` crate

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

    No

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -1,3 +1,3 @@
 [workspace]
 
-members = ["xline", "curp", "benchmark", "utils", "engine", "xlineapi"]
+members = ["xline", "curp", "benchmark", "utils", "engine", "xlineapi", "curp-extapi", "curp-testutils"]
```
Comment (Phoenix500526) [inline Cargo.toml:3]: I think curp-external-api would be more readable.

File: curp-extapi/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "curp-extapi"
```
Comment (Phoenix500526) [inline curp-extapi/Cargo.toml:2]: Same as above

File: curp-testutils/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "curp-testutils"
```
Comment (Phoenix500526) [inline curp-testutils/Cargo.toml:2]: I recommend `curp-test-utils` since utils is an initialism.

File: curp/src/lib.rs [hunk 4]
```diff
@@ -180,5 +184,5 @@ mod rpc;
 /// Snapshot
 mod snapshot;
 
-#[cfg(test)]
-pub(crate) mod test_utils;
+// #[cfg(test)]
```
Comment (Phoenix500526) [inline curp/src/lib.rs:187]: Plz remove these useless comments if we don't need it anymore.

File: curp-testutils/src/test_cmd.rs [hunk 5]
```diff
@@ -43,15 +43,29 @@ impl Display for ExecuteError {
 
 #[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
 pub struct TestCommand {
-    id: ProposeId,
-    keys: Vec<u32>,
+    pub id: ProposeId,
+    pub keys: Vec<u32>,
```
Comment (rogercloud) [inline curp-testutils/src/test_cmd.rs:47]: Can we provide functions to access these field?

File: Cargo.toml (hunk 1 above)
Comment (bsbds) [reply Cargo.toml]: done

File: curp-extapi/Cargo.toml (hunk 2 above)
Comment (bsbds) [reply curp-extapi/Cargo.toml]: done

File: curp-testutils/Cargo.toml (hunk 3 above)
Comment (bsbds) [reply curp-testutils/Cargo.toml]: done

File: curp/src/lib.rs (hunk 4 above)
Comment (bsbds) [reply curp/src/lib.rs]: done

File: curp-testutils/src/test_cmd.rs (hunk 5 above)
Comment (bsbds) [reply curp-testutils/src/test_cmd.rs]: Sorry the `pub` is actually not needed as the public `Commad` trait provides these get funtions. 

Fixed.

File: curp/tests/common/mod.rs [hunk 6]
```diff
@@ -1,82 +1,3 @@
 #![allow(dead_code, unused)]
```
Comment (themanforfree) [inline curp/tests/common/mod.rs:1]: Please remove this

File: curp-external-api/src/log_index.rs [hunk 7]
```diff
@@ -0,0 +1,2 @@
+/// Log Index
+pub type LogIndex = u64;
```
Comment (themanforfree) [inline curp-external-api/src/log_index.rs:2]: I recommend define this type in other file and remove this file

File: curp-external-api/src/role_change.rs [hunk 8]
```diff
@@ -1,8 +1,9 @@
-#[cfg(test)]
 use mockall::automock;
 
 /// Callback when the leadership changes
-#[cfg_attr(test, automock)]
+#[allow(clippy::indexing_slicing)]
+#[allow(clippy::integer_arithmetic)]
```
Comment (themanforfree) [inline curp-external-api/src/role_change.rs:5]: These two lines seem to be meaningless

**Comment by bsbds** [comment]
> Please draw a diagram to describe the dependencies among these three crates and update it to the relevant issue.

Updated in https://github.com/datenlord/Xline/issues/290#issuecomment-1583943701

File: curp/src/server/cmd_worker/mod.rs [hunk 9]
```diff
@@ -305,7 +303,7 @@ mod tests {
         let cmd = Arc::new(TestCommand::default());
 
         ce_event_tx.send_sp_exe(Arc::clone(&cmd), 1);
-        assert_eq!(er_rx.recv().await.unwrap().1, vec![]);
+        assert_eq!(er_rx.recv().await.unwrap().1 .1, vec![]);
```
[the remark this answers is not in the document]
Comment (bsbds) [reply curp/src/server/cmd_worker/mod.rs:306]: The `Vec<u32>` is the value and `Vec<i64>` is the revision. The type of `TestCommandResult` came from the curp integration tests utils, the integration tests need the revision result. 

A comment is added to `TestCommandResult`.

File: curp/tests/common/mod.rs (hunk 6 above)
Comment (bsbds) [reply curp/tests/common/mod.rs:1]: The integration tests contains some unused code, this lint is needed

File: curp-external-api/src/log_index.rs (hunk 7 above)
Comment (bsbds) [reply curp-external-api/src/log_index.rs:2]: You mean move it to the `lib.rs` file?

Comment (themanforfree) [reply curp-external-api/src/log_index.rs:2]: Yes

**Review by Phoenix500526 — APPROVED** [review]

**Review by themanforfree — APPROVED** [review]

**Pull request merged**